### PR TITLE
Task 5

### PR DESCRIPTION
**Pull request by noway36**
Total: 8

import-service was created with it's own sls config. It returns Signed Url by GET request 
importProductsFile implemented (`name` marked as query string request parameter)
frontend integrated with new import service [FE PR link](https://github.com/noway36/shop-react-redux-cloudfront/pull/3)
[APP](https://d1mj375gxygqlt.cloudfront.net/admin/products)

importFileParser implemented and it is triggered by the S3 s3:ObjectCreated:* event.
importFileParser uses Readable stream to get an object from S3 and csv-parser to parse it

Additional (optional) tasks:

- async/await is used in lambda functions
- importProductsFile lambda is covered by unit tests
- At the end of the stream the lambda function moves the file from the uploaded folder into the parsed folder

In case of any questions, please, contact Aliaksandr Dren in Teams. 🙂

### CODE REVIEW

File: import-service/serverless.ts [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+    //                         {
+    //                             AllowedMethods: ['GET', 'PUT'],
+    //                             AllowedHeaders: ['*'],
+    //                             AllowedOrigins: ['*'],
+    //                         },
+    //                     ],
+    //                 },
+    //             },
+    //         },
+    //     },
```
Comment (lyolka92) [inline import-service/serverless.ts:64]: I recommend you to return the Resources section and add s3BucketPolicy there.

File: import-service/src/S3Service.ts [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+                throw error;
+            });
+    }
+
+    private async copyObject(event: S3Event, from: CsvFileFolder, to: CsvFileFolder) {
+        await this.s3
+            .copyObject({
+                Bucket: BUCKET,
+                CopySource: `${BUCKET}/${this.extractKeyFromEvent(event)}`,
+                Key: this.extractKeyFromEvent(event).replace(from, to),
```
Comment (lyolka92) [inline import-service/src/S3Service.ts:56]: What if your file name will include the word "uploaded"?

File: import-service/src/functions/importFileParser/index.ts [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+export default {
+    handler: `${handlerPath(__dirname)}/handler.main`,
+    events: [
+        {
+            s3: {
+                bucket: BUCKET,
+                existing: true,
+                rules: [{ prefix: `${CsvFileFolder.Uploaded}/` }],
+                event: 's3:ObjectCreated:*',
+                suffix: '.csv',
```
Comment (lyolka92) [inline import-service/src/functions/importFileParser/index.ts]: As I know, 'suffix' should be in the 'rules' section.

File: import-service/src/S3Service.ts (hunk 2 above)
Comment (noway36) [reply import-service/src/S3Service.ts:56]: I've tried to import `uploaded.csv` and `uploaded_test.csv` and there were no problems
image

Is there any case I don't know about?

File: import-service/src/functions/importFileParser/index.ts (hunk 3 above)
Comment (noway36) [reply import-service/src/functions/importFileParser/index.ts]: Ye, it seems you're right, thanks

File: import-service/serverless.ts (hunk 1 above)
Comment (noway36) [reply import-service/serverless.ts:64]: I have a problem with deploying it. `Bucket already exists` error, or something like that, so I decided to comment Resources section. 
If you know how to deal with this case, please, let me know

Comment (lyolka92) [reply import-service/serverless.ts:64]: I had the same problem. Try to remove the existing bucket and then run `sls deploy`.

**Pull request merged**